### ISSUE TITLE
backend/drm: figure out commit mechanism

### ISSUE THREAD

**Issue by emersion** [swaywm/wlroots#2239]
Right now in the DRM backend all commits that aren't page-flips are blocking. This isn't great because this potentially stalls the compositor for a long time.

We should aim to make all atomic commits non-blocking instead.

However this would break compositors which perform a modeset while waiting for a page-flip event. KMS would return EBUSY in this case. This is because KMS doesn't implement a "mailbox" model: with KMS, 1 commit = 1 page-flip and you can't "amend" the last commit.

Here are possible solutions.

## Make compositors responsible for accumulating state

This would be cumbersome for compositors, they'd need to wait for the vblank event before performing a modeset. But this would continue to make our backend interface the least common denominator and also wouldn't introduce more magic (which has proven to be a mistake in the past, see e.g. https://github.com/swaywm/wlroots/issues/1363 and https://github.com/swaywm/wlroots/issues/2094).

In practice, compositors would dp a test-only modeset with `wlr_output_test`, stash the new mode, and apply it on next vblank event.

## Emulate a mailbox model on top of KMS

Basically, allow compositors to do as many commits as they want, accumulate them in a pending state, and only submit the last pending state to KMS right before vblank. This is simpler for compositors and would allow us to make `wlr_output` more like `wl_surface`.

But what does "right before vblank" mean? We'll need a timer and a reasonable delay. We'll potentially break compositors that already use a timer to submit a frame as late as possible.

## Implement DRM_MODE_ATOMIC_AMEND in the kernel

This has been suggested in the past. Would be more work and even when it's merged we'd need to wait quite a while before being able to use it unconditionally (maybe keep blocking commits if not supported?). But that would be the nicest solution.

cc @ascent12
